pkg/auth/auth.go

```diff
@@ -0,0 +1,184 @@
+package auth
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"path"
+	"strings"
+)
+
+// ConfigEntry - Configuration for authentication
+type ConfigEntry struct {
+	Type    string `yaml:"type"`
+	Enabled bool   `yaml:"enabled"`
+}
+
+// Provider - an auth provider is an adapter that provides the principal
+// object required for authentication. This can be a User, a System, or some
+// other entity.
+type Provider interface {
+	GetPrincipal(*http.Request) (Principal, error)
+}
+
+// Principal - principal user or other identity of some kind with access to the
+// broker.
+type Principal interface {
+	GetType() string
+	GetName() string
+	// TODO: add roles?
+}
+
+// UserServiceAdapter - is the interface for a service that stores Users. It can
+// be anything you want: file, database, whatever as long as you can search and
+// validate them.
+type UserServiceAdapter interface {
+	FindByLogin(string) (User, error)
+	ValidateUser(string, string) bool
+}
+
+// User - a User from the service adapter
+type User struct {
+	Username string
+	Password string
+}
+
+// GetType - returns the type of Principal. This is a user principal.
+func (u User) GetType() string {
+	return "user"
+}
+
+// GetName - returns the name of the User Principal.
+func (u User) GetName() string {
+	return u.Username
+}
+
+// FileUserServiceAdapter - a file based UserServiceAdapter which seeds its
+// users from a file.
+type FileUserServiceAdapter struct {
+	filedir string
+	userdb  map[string]User
+}
+
+// NewFileUserServiceAdapter - constructor for the FUSA
+func NewFileUserServiceAdapter(dir string) (*FileUserServiceAdapter, error) {
+	if dir == "" {
+		dir = "/var/run/asb-auth"
+	}
+	fusa := FileUserServiceAdapter{filedir: dir}
+	err := fusa.buildDb()
+	if err != nil {
+		return nil, err
+	}
+	return &fusa, nil
+}
+
+func (d *FileUserServiceAdapter) buildDb() error {
+	userfile := path.Join(d.filedir, "username")
+	passfile := path.Join(d.filedir, "password")
+	username, uerr := ioutil.ReadFile(userfile)
+	if uerr != nil {
+		return uerr
+	}
+	password, perr := ioutil.ReadFile(passfile)
+	if perr != nil {
+		return perr
+	}
+
+	// userdb is probably overkill, but if we ever want to allow multiple users,
+	// it'll come in handy.
+	d.userdb = make(map[string]User)
+	unamestr := string(username)
+	d.userdb[unamestr] = User{Username: unamestr, Password: string(password)}
+
+	return nil
+}
+
+// FindByLogin - given a login name, this will return the associated User or an
+// error
+func (d FileUserServiceAdapter) FindByLogin(login string) (User, error) {
+
+	// TODO: add some error checking
+	if user, ok := d.userdb[login]; ok {
+		return user, nil
+	}
+
+	return User{}, errors.New("user not found")
+
+}
+
+// ValidateUser - returns true if the given user's credentials match a user in
+// our backend storage. Returns fals otherwise.
+func (d FileUserServiceAdapter) ValidateUser(username string, password string) bool {
+	user, err := d.FindByLogin(username)
+	if err != nil {
```

Comment:
Probably, I don't have the log in these structs so I removed the fmt I had. Definitely will log though. Thanks.